bootstrap.md

```diff
@@ -0,0 +1,136 @@
+
+A strawman bootstrapping sequence would have nodes use these initial peers to sync up to the canonical chain. However, new connection requests might effecitvely DoS these peers, making the Filecoin network unavailable to new nodes. Therefore, Filecoin bootstrapping seeks to "load-balance" initial requests to ensure the network remains widely available.
+
+We connect to each of these initial peers using the `hello handshake` (see in [network protocols](./network-protocols.md#hello-handshake)), getting back
+
+```go
+type HelloMessage struct {
+HeaviestTipSet []Cid
+HeaviestTipSetHeight uint64
+GenesisHash Cid
+}
+```
+
+followed by a `GetPeers` request, getting back
+
+```go
+type PeersMessage struct {
+PeerSet []PeerID
+}
+```
+
+We will use these new peers to sync with the chain, using the chain heads from our initial set of trusted nodes as a means of verifying that these peers are mining the canonical chain. These initial peers do not return their latest heads, but rather their latest confirmed heads, thereby ensuring other honest peers should be aware of these as well.
+
+We assume no long-lived network partitions in Filecoin, meaning these confirmed heads will all be on the same chain (though potentially at different heights). A node should initially connect to 5 trusted nodes to begin bootstrapping.
+
+```go
+type Bootstrapper struct {
+    trustedPeers []PeerID
+    trustedHeads map[[]Cid]uint64
+    genesisHash Cid
+    
+    bootstrapPeers []PeerID
+    
+}
+
+func InitialConnect(config ConfigFile, b Bootstrapper, s Syncer) {
+    b.trustedPeers = configFile.bootstrapPeers
+    
+    for trustedPeer := range b.trustedPeers {
+        trustedChain := sayHello(trustedPeer)
+        otherPeers := getPeers(trustedPeer).PeerSet
+        
+        b.trustedHeads[trustedChain.HeaviestTipSet] = trustedChain.HeaviestTipSetHeight
+    	
+        if b.genesisHash == nil {
+	        b.genesisHash = trustedChain.GenesisHash           
+        } else if b.genesisHash != trustedChain.GenesisHash {
+            panic("One of your trusted peers is bad.")
+        }
+        
+        b.bootstrapPeers = append(b.bootstrapPeers, otherPeers)
+    }
+    
+    s.genesis = b.genesisHash
+}
+```
+
+
+
+## Peer Set Expansion
+
+Once connected to some portion of the network, a node will need to expand their peer set in order to get a good sample of the network and a live latest head. To do so, use the DHT to do random `FindPeer` requests. We do so until we have requested 5 new peers from each peer in the PeerSet.
+
+For each of our 30 peers, we perform hello handshakes, and run the following verifications:
+
+- Ensure that each of our bootstrapper's `trustedHeads.keys` are included as ancestors of our peers' latestHeads.
+- If they are not, include that peer in our syncer's `BadTipsetCache`.
+- If they are, add these peers and latestHeads to our syncer's peerSet/peerHeads (with offset initialConnections)
```

Comment:
I can fill the above in with pseudocode, provided we agree on this being a good process and way the Syncer is built up by bootstrapping

Reply:
where does trustedHeads come from?

Reply:
Ship from a config file with Filecoin nodes. Can be changed manually by clients running non-vanilla filecoin (eg mining pools or the like).